pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s together with a parameter update triggering '%s' is not allowed", old.Namespace, old.Name, newOvRef, *triggeredPlan)
+	case isParameterUpdate && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: triggering one plan '%s' directly and through parameter update '%s' is not allowed", old.Namespace, old.Name, oldPlan, newPlan)
+	case isParameterUpdate && isPlanOverride:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: updating parameters and triggering plan '%s' is not allowed", old.Namespace, old.Name, *triggeredPlan)
+	}
+
+	// Deciding which plan to trigger:
+	switch {
+	case isUpgrade:
+		plan := SelectPlan([]string{UpgradePlanName, UpdatePlanName, DeployPlanName}, ov)
+		if plan == nil {
+			return nil, fmt.Errorf("failed to update Instance %s/%s: couldn't find any suitable plan that would be triggered by an OperatorVersion upgrade", old.Namespace, old.Name)
+		}
+		return plan, nil
+
+	case isParameterUpdate:
+		// if the same plan is triggered by the update, we clean the Instance.Status to effectively restart the plan
+		if *triggeredPlan == oldPlan {
+			if err := resetInstanceStatus(old); err != nil {
+				return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
+			}
+		}
+		return triggeredPlan, nil
+
+	case isNovelPlan:
+		return &newPlan, nil
+		// Implement plan overrides and cancellations below:
+	}
+
+	return triggeredPlan, nil
+}
+
+// resetInstanceStatus clears Instance.Status to effectively restart existing plan
+func resetInstanceStatus(instance *Instance) error {
+	// TODO (AD): implement
+	return nil
+}
+
+// getOperatorVersion retrieves operator version belonging to the given instance
+func (ia *InstanceAdmission) getOperatorVersion(instance *Instance) (ov *OperatorVersion, err error) {
+	ov = &OperatorVersion{}
+	err = ia.client.Get(context.TODO(),
+		types.NamespacedName{
+			Name:      instance.Spec.OperatorVersion.Name,
+			Namespace: instance.OperatorVersionNamespace(),
+		},
+		ov)
+	if err != nil {
+		log.Printf("InstanceAdmission: Error getting operatorVersion %s for instance %s: %v", instance.Spec.OperatorVersion.Name, instance.Name, err)
+		return nil, err
+	}
+	return ov, nil
+}
+
+// triggeredPlan determines what plan to run based on parameters that changed and the corresponding parameter trigger.
+func triggeredPlan(params []Parameter, ov *OperatorVersion) (*string, error) {
+	// If no parameters were changed, we return an empty string so no plan would be triggered
+	if len(params) == 0 {
+		return nil, nil
+	}
+
+	plans := make([]string, 0)
+	for _, p := range params {
+		if p.Trigger != "" && SelectPlan([]string{p.Trigger}, ov) != nil {
+			plans = append(plans, p.Trigger)
+		}
+	}
+
+	plans = funk.UniqString(plans)
+
+	switch len(plans) {
+	case 0:
+		// no plan could be triggered since we do not force existence of the "deploy" plan in the operators
```

Comment:
I don't understand this comment 🤔

Reply:
Funny enough, you wrote it 😂 (I copied it from one of your first commits). Basically, we're saying that the whole `case 0` and searching for a `fallback` is possible/needed because we don't enforce the existence of the `deploy` plan in the operator.

Reply:
Really? Well then that's a perfect comment, let's keep it :D 😂